src/main.rs

```diff
@@ -96,3 +96,169 @@ fn main() {
         messenger.sender(),
     );
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::interfaces::connection::ConnectionService;
+    use crate::interfaces::messenger::Messenger;
+    use crate::interfaces::packet_processor::PacketProcessor;
+    use crate::models::minecraft_protocol::MinecraftProtocolReader;
+    use crate::*;
+
+    fn start_trace() {
+        let logger_config = ConfigBuilder::new()
+            .set_max_level(LevelFilter::Off)
+            .set_thread_level(LevelFilter::Off)
+            .set_target_level(LevelFilter::Off)
+            .build();
+        SimpleLogger::init(LevelFilter::Trace, logger_config).unwrap();
+    }
+
+    fn handle_connection<M: Messenger, PP: PacketProcessor, F: Fn()>(
+        mut stream: std::net::TcpStream,
+        inbound_packet_processor: PP,
+        messenger: M,
+        id: uuid::Uuid,
+        on_closure: F,
+        sx: std::sync::mpsc::Sender<i32>,
+    ) {
+        let stream_clone = stream.try_clone().expect("Failed to clone stream");
+        messenger.new_connection(id, stream_clone);
+        loop {
+            match stream.try_read_var_int() {
+                Ok(length) => {
+                    sx.send(length).expect("Failed to send data to channel");
+                }
+                Err(e) => {
+                    match e.kind() {
+                        std::io::ErrorKind::UnexpectedEof => on_closure(),
+                        std::io::ErrorKind::ConnectionReset => on_closure(),
+                        _ => {
+                            panic!("Connection closed due to {:?}", e);
+                        }
+                    };
+                    break;
+                }
+            }
+        }
+    }
```

Comment:
Why are we duplicating code from the server module here?

Reply:
I duplicated the code here because I did not find any other way to send information about received packets to the test thread (main thread). What has changed is line 130, where I send (for the moment) the length of the packet that has been received.

Reply:
Since the server creates a new thread to handle connection, I needed a way to send/receive data. Hence I needed a way to pass in channels to the handle_connection() function

Reply:
I think that duplicating the code has two main issues:
- More work whenever we need to update something
- The test no longer tests the software, but a copy of the software that can become out of date

I'm fine with there being some duplication but my gist from reading this was that you literally copy pasted a ton of code into this method (given the log lines which you probably didn't rewrite for this test). Can you do a pass where you try to eliminate as much code replication as possible? You can refactor the non-test code as needed to make it work better, but I don't think this code is merge-able as-is

Reply:
Ok, I'll rework that then